solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -80,6 +80,7 @@
 
 public class ZkStateReader implements SolrCloseable {
   public static final int STATE_UPDATE_DELAY = Integer.getInteger("solr.OverseerStateUpdateDelay", 2000);  // delay between cloud state updates
+  public volatile PerReplicaStateUpdateListener testPerReplicaStateUpdateListener = null;
```

Comment:
Perhaps we can move this to `PerReplicaStatesOps` as a `private static volatile PerReplicaStateUpdateListener listener;` add a public setter method there, then we don't need to modify the method signature of `PerReplicaStatesOps#persist` (and all the callers 😊 ). In `PerReplicaStatesOps#persist` we can simply check `listener != null` as you have done similarly in `ClusterStateUpdater` with `testOverseerMessageListener` ?

We can then set it from `FSPRSTest#testPRSCollectionCreation` instead, which makes it default as null for normal execution flow? 😄

Reply:
I think ZkStateReader is singleton class at coreContainer level. that's why I kept variable at that level. I didn't want static variable as we create multiple coreContainer in test.

Reply:
static variable is also guaranteed a singleton (within the same classloader - well same as ZkStateReader singleton anyway)and it's closer to where it's used (PerReplicaStateOps), which method `persist` is defined.

The major benefit of that is it avoids any method signature change, hence we do not have to change any of the 5 invocations 😊

Reply:
ah i see what u mean - since in ur test cases u started 2 `JettySolrRunner` so u r setting it per runner/coreContainer. Hmm lemme see...

Reply:
1. Let's keep ZkClient for separate PR if require (Or we can use zkversion of node in some cases ).  
2. I'm not sure it is using classloader. I think it just creating different instances
3.  ok, persist api issue. This is api argument to listen the events. Unfortunately,  I'm not finding any better way to solve this issue as PerReplicaStatesOps instance  is created for each update.

Reply:
Putting such listener as a part of ZkClient should solve 3. here? TBH, ZkClient has already provided whole bunch of other "info" so adding a PRS op listener shouldn't be too bad? 😅 
![image](https://user-images.githubusercontent.com/2895902/187545822-863509ae-3e79-417c-b153-674a4335ef92.png)